src/particles/diagnostics/ReducedBeamCharacteristics.cpp

```diff
@@ -0,0 +1,224 @@
+/* Copyright 2023 The Regents of the University of California, through Lawrence
+ *           Berkeley National Laboratory (subject to receipt of any required
+ *           approvals from the U.S. Dept. of Energy). All rights reserved.
+ *
+ * This file is part of ImpactX.
+ *
+ * Authors: Marco Garten, Yinjian Zhao
+ * License: BSD-3-Clause-LBNL
+ */
+
+#include "ImpactX.H"
+#include "particles/ImpactXParticleContainer.H"
+#include "particles/diagnostics/ReducedBeamCharacteristics.H"
+
+#include <AMReX_ParallelDescriptor.H>   // for ParallelDescriptor
+#include <AMReX_REAL.H>                 // for ParticleReal
+#include <AMReX_Reduce.H>               // for ReduceOps
+#include <AMReX_ParticleReduce.H>       // for ParticleReduce
+#include <AMReX_ParallelDescriptor.H>   // for ParallelDescriptor
+#include <AMReX_BLProfiler.H>           // for TinyProfiler
+
+
+namespace impactx
+{
+
+    namespace diagnostics
+    {
+
+        std::unordered_map<std::string, amrex::ParticleReal> reduced_beam_characteristics(ImpactXParticleContainer const & pc) {
+
+            BL_PROFILE("impactX::diagnostics::reduced_beam_characteristics");
+
+            // preparing to access reference particle data: RefPart
+            RefPart const ref_part = pc.GetRefParticle();
+            // reference particle charge in units of positive elementary charge
+            amrex::ParticleReal const q_qe = ref_part.charge_qe();
+
+            // preparing access to particle data: AoS and SoA
+            using PTDType = typename ImpactXParticleContainer::ParticleTileType::ConstParticleTileDataType;
+
+            amrex::ReduceOps<amrex::ReduceOpSum,amrex::ReduceOpSum,amrex::ReduceOpSum,amrex::ReduceOpSum,amrex::ReduceOpSum,
+            amrex::ReduceOpSum,amrex::ReduceOpSum> reduce_ops;
+
+            auto r = amrex::ParticleReduce<amrex::ReduceData<amrex::ParticleReal,amrex::ParticleReal,
+            amrex::ParticleReal,amrex::ParticleReal,amrex::ParticleReal,amrex::ParticleReal,amrex::ParticleReal>>(
+                    pc,
+                    [=] AMREX_GPU_DEVICE(const PTDType& ptd, const int i) noexcept -> amrex::GpuTuple
+                            <amrex::ParticleReal,amrex::ParticleReal,amrex::ParticleReal,amrex::ParticleReal,amrex::ParticleReal,
+                            amrex::ParticleReal,amrex::ParticleReal>
+                    {
+                        // access AoS particle position data
+                        const amrex::ParticleReal p_pos0 = ptd.m_aos[i].pos(0);
+                        const amrex::ParticleReal p_pos1 = ptd.m_aos[i].pos(1);
+                        const amrex::ParticleReal p_pos2 = ptd.m_aos[i].pos(2);
```

Comment:
Unless I am mistaken, you cannot access `ptd.m_aos` as a whole here.
Use the pattern we use in WarpX for now please.
https://github.com/ECP-WarpX/WarpX/blob/872bae2026a6767f2023f8ba600c87d247eed38e/Source/Diagnostics/ReducedDiags/BeamRelevant.cpp#L198-L209

Reply:
> Unless I am mistaken, you cannot access ptd.m_aos as a whole here.   

Sorry, could you explain? 🤔  It works on MacOS and on Perlmutter with all tests that I did. 
I was trying the pattern you mentioned before but the one I am using now is the one that @atmyers suggested.

Reply:
As discussed on Slack: could be a compiler bug for the NVCC version in CI. Please try again w/ syntax in the link above.